paddle/math/float16.h

```diff
@@ -0,0 +1,714 @@
+}
+
+__device__ inline float16 operator/(const float16& a, const float16& b) {
+  // TODO(kexinzhao): check the cuda version that starts to support __hdiv
+  float num = __half2float(half(a));
+  float denom = __half2float(half(b));
+  return float16(num / denom);
+}
+
+__device__ inline float16 operator-(const float16& a) {
+  return float16(__hneg(half(a)));
+}
+
+__device__ inline float16& operator+=(float16& a, const float16& b) {
+  a = a + b;
+  return a;
+}
+
+__device__ inline float16& operator-=(float16& a, const float16& b) {
+  a = a - b;
+  return a;
+}
+
+__device__ inline float16& operator*=(float16& a, const float16& b) {
+  a = a * b;
+  return a;
+}
+
+__device__ inline float16& operator/=(float16& a, const float16& b) {
+  a = a / b;
+  return a;
+}
+
+__device__ inline bool operator==(const float16& a, const float16& b) {
+  return __heq(half(a), half(b));
+}
+
+__device__ inline bool operator!=(const float16& a, const float16& b) {
+  return __hne(half(a), half(b));
+}
+
+__device__ inline bool operator<(const float16& a, const float16& b) {
+  return __hlt(half(a), half(b));
+}
+
+__device__ inline bool operator<=(const float16& a, const float16& b) {
+  return __hle(half(a), half(b));
+}
+
+__device__ inline bool operator>(const float16& a, const float16& b) {
+  return __hgt(half(a), half(b));
+}
+
+__device__ inline bool operator>=(const float16& a, const float16& b) {
+  return __hge(half(a), half(b));
+}
+
+// On ARMv8.2-A CPU
+#elif defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
+    (PADDLE_GNUC_VER >= 71 || PADDLE_CLANG_VER >= 39)
+__host__ inline float16 operator+(const float16& a, const float16& b) {
+  return float16(vaddh_f16(float16_t(a), float16_t(b)));
+}
+
+__host__ inline float16 operator-(const float16& a, const float16& b) {
+  return float16(vsubh_f16(float16_t(a), float16_t(b)));
+}
+
+__host__ inline float16 operator*(const float16& a, const float16& b) {
+  return float16(vmulh_f16(float16_t(a), float16_t(b)));
+}
+
+__host__ inline float16 operator/(const float16& a, const float16& b) {
+  return float16(vdivh_f16(float16_t(a), float16_t(b)));
+}
+
+__host__ inline float16 operator-(const float16& a) {
+  return float16(vnegh_f16(float16_t(a)));
+}
+
+__host__ inline float16& operator+=(float16& a, const float16& b) {
+  a = a + b;
+  return a;
+}
+
+__host__ inline float16& operator-=(float16& a, const float16& b) {
+  a = a - b;
+  return a;
+}
+
+__host__ inline float16& operator*=(float16& a, const float16& b) {
+  a = a * b;
+  return a;
+}
+
+__host__ inline float16& operator/=(float16& a, const float16& b) {
+  a = a / b;
+  return a;
+}
+
+__host__ inline bool operator==(const float16& a, const float16& b) {
+  return static_cast<bool>(vceqh_f16(float16_t(a), float16_t(b)));
+}
+
+__host__ inline bool operator!=(const float16& a, const float16& b) {
+  return !(a == b);
+}
+
+__host__ inline bool operator<(const float16& a, const float16& b) {
+#ifdef PADDLE_NEON_64
```

Comment:
Is still need `PADDLE_NEON_64` here?
This code is under the macro of `PADDLE_ARM_FP16`.

Reply:
PADDLE_ARM_FP16 is CPU Architecture related (intend to be defined when ARMv8.2A arch is found). 
PADDLE_NEON_64 is more about the execution state of ARMv8.2A, because I believe ARMv8.2A CPU can run either in 32bit (when __arm__ is defined) or 64 bit (when __aarch64__ is defined). GCC provides different sets of ARM intrinsics for __arm__ and __aarch64__. That's why I define PADDLE_NEON_64 here.

Reply:
So, when I specify the architecture as ARMv8.2(for those float16 instructions), can I still compile a 32-bit program?

Reply:
I don't think so.  

An ARMv8 cpu only runs on `arm`-32bit state when the operating system is 32bit, which is the case for Raspberry Pi 3 model B.

I don't think anyone would run a 32bit OS on a ARMv8.2 cpu. So I will delete PADDLE_NEON_64

Reply:
A 32bit OS can run on an ARMv8.2 cpu. Also, a 32bit program can run on a 64bit OS(on ARMv8.2 cpu).
My point is, when you compile a program that uses the float16 instruction, it may only be compiled into a 64-bit program.

Reply:
To reflect this, current code assumes 64-bit compilation when PADDLE_ARM_FP16 is defined.